src/opf/acp.jl

```diff
@@ -1,293 +1,167 @@
     T=Float64,    
 )
-    # Cleanup and pre-process data
-    PM.standardize_cost_terms!(data, order=2)
-    PM.calc_thermal_limits!(data)
-    ref = PM.build_ref(data)[:it][:pm][:nw][0]
-
-    # Check that slack bus is unique
-    length(ref[:ref_buses]) == 1 || error("Data has $(length(ref[:ref_buses])) slack buses (expected 1).")
-    i0 = first(keys(ref[:ref_buses]))
+    # TODO: remove when all formulations are done
+    data = OPFData(network)
     
     # Grab some data
-    N = length(ref[:bus])
-    G = length(ref[:gen])
-    E = length(data["branch"])
-    L = length(ref[:load])
-    bus_loads = [
-        [ref[:load][l] for l in ref[:bus_loads][i]]
-        for i in 1:N
-    ]
-    bus_shunts = [
-        [ref[:shunt][s] for s in ref[:bus_shunts][i]]
-        for i in 1:N
-    ]
+    N, E, G = data.N, data.E, data.G
+    vmin, vmax = data.vmin, data.vmax
+    i0 = data.ref_bus
+    gs, bs = data.gs, data.bs
+    pd, qd = data.pd, data.qd
+    bus_arcs_fr, bus_arcs_to = data.bus_arcs_fr, data.bus_arcs_to
+    bus_gens = data.bus_gens
+    pgmin, pgmax = data.pgmin, data.pgmax
+    qgmin, qgmax = data.qgmin, data.qgmax
+    c0, c1, c2 = data.c0, data.c1, data.c2
+    gen_status = data.gen_status
+    bus_fr, bus_to = data.bus_fr, data.bus_to
+    gff, gft, gtf, gtt = data.gff, data.gft, data.gtf, data.gtt
+    bff, bft, btf, btt = data.bff, data.bft, data.btf, data.btt
+    dvamin, dvamax, smax = data.dvamin, data.dvamax, data.smax
+    branch_status = data.branch_status
 
     model = JuMP.GenericModel{T}(optimizer)
-    model.ext[:opf_model] = ACOPF  # for internal checks
+    model.ext[:opf_model] = ACOPF
 
     #
     #   I. Variables
     #
+    
     # nodal voltage
-    JuMP.@variable(model, va[1:N])
-    JuMP.@variable(model, ref[:bus][i]["vmin"] <= vm[i in 1:N] <= ref[:bus][i]["vmax"], start=1.0)
+    @variable(model, vm[i in 1:N], start=1.0)
+    @variable(model, va[1:N])
+
     # Active and reactive dispatch
-    JuMP.@variable(model, ref[:gen][g]["pmin"] <= pg[g in 1:G] <= ref[:gen][g]["pmax"])
-    JuMP.@variable(model, ref[:gen][g]["qmin"] <= qg[g in 1:G] <= ref[:gen][g]["qmax"])
-    # Bi-directional branch flows
-    JuMP.@variable(model, -ref[:branch][l]["rate_a"] <= pf[(l,i,j) in ref[:arcs]] <= ref[:branch][l]["rate_a"])
-    JuMP.@variable(model, -ref[:branch][l]["rate_a"] <= qf[(l,i,j) in ref[:arcs]] <= ref[:branch][l]["rate_a"])
+    @variable(model, pg[g in 1:G])
+    @variable(model, qg[g in 1:G])
+
+    # Directional branch flows
+    @variable(model, pf[e in 1:E])
+    @variable(model, qf[e in 1:E])
+    @variable(model, pt[e in 1:E])
+    @variable(model, qt[e in 1:E])
 
     # 
     #   II. Constraints
     #
+
+    # Voltage magnitude bounds
+    set_lower_bound.(vm, vmin)
+    set_upper_bound.(vm, vmax)
+
+    # Active generation bounds (both zero if generator is off)
+    set_lower_bound.(pg, gen_status .* pgmin)
+    set_upper_bound.(pg, gen_status .* pgmax)
```

Comment:
This will set lower/upper bound to zero if `gen_status` is zero.

I can see this potentially becoming an issue for Ipopt, which tends to not like having variables with same lower/upper bounds. Does not need to be changed, just to keep in mind if we see Ipopt having convergence issues when removing generators/branches